File: packages/ios-sdk/Sources/OpencomSDK/Core/OpencomAPIClient.swift
import Foundation
import UIKit

/// HTTP client for communicating with the Convex backend.
actor OpencomAPIClient {
    private let config: OpencomConfig
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    init(config: OpencomConfig) {
        self.config = config
        self.session = URLSession.shared
        self.decoder = JSONDecoder()
        self.encoder = JSONEncoder()
    }

    // MARK: - Session APIs

    func bootSession(
        device: DeviceInfo
    ) async throws -> BootSessionResponse {
        let body: [String: Any] = [
            "path": "widgetSessions:boot",
            "args": [
                "workspaceId": config.workspaceId,
                "device": device.toDictionary()
            ]
        ]

        return try await post(body: body)
    }

    func refreshSession(sessionToken: String) async throws -> RefreshSessionResponse {
        let body: [String: Any] = [
            "path": "widgetSessions:refresh",
            "args": [
                "workspaceId": config.workspaceId,
                "sessionToken": sessionToken
            ]
        ]

        return try await post(body: body)
    }

    func revokeSession(sessionToken: String) async throws {
        let body: [String: Any] = [
            "path": "widgetSessions:revoke",
            "args": [
                "workspaceId": config.workspaceId,
                "sessionToken": sessionToken
            ]
        ]

        let _: EmptyResponse = try await post(body: body)
    }

    // MARK: - Visitor APIs

    func identifyVisitor(
        visitorId: String,
        sessionToken: String?,
        user: OpencomUser,
        device: DeviceInfo
    ) async throws {
        var args: [String: Any] = [
            "visitorId": visitorId,
            "device": device.toDictionary()
        ]

        if let sessionToken = sessionToken {
            args["sessionToken"] = sessionToken
        }

        if let email = user.email {
            args["email"] = email
        }
        if let name = user.name {
            args["name"] = name
        }
        if let userId = user.userId {
            args["externalUserId"] = userId
        }

        var customAttributes: [String: Any] = [:]
        if let company = user.company {
            customAttributes["company"] = company
        }
        if let attrs = user.customAttributes {
            for (key, value) in attrs {
                customAttributes[key] = value
            }
        }
        if !customAttributes.isEmpty {
            args["customAttributes"] = customAttributes
        }

        let body: [String: Any] = [
            "path": "visitors:identify",
            "args": args
        ]

        let _: EmptyResponse = try await post(body: body)
    }

    func heartbeat(visitorId: String, sessionToken: String?) async throws {
        var args: [String: Any] = ["visitorId": visitorId]
        if let sessionToken = sessionToken {
            args["sessionToken"] = sessionToken
        }
        let body: [String: Any] = [
            "path": "visitors:heartbeat",
            "args": args
        ]

        let _: EmptyResponse = try await post(body: body)
    }

    // MARK: - Event APIs

    func trackEvent(
        visitorId: String,
        sessionToken: String?,
        name: String,
        properties: [String: Any]?,
        sessionId: String
    ) async throws {
        var args: [String: Any] = [
            "visitorId": visitorId,
            "name": name,
            "sessionId": sessionId
        ]

        if let sessionToken = sessionToken {
            args["sessionToken"] = sessionToken
        }

        if let properties = properties {
            args["properties"] = properties
        }

        let body: [String: Any] = [
            "path": "events:track",
            "args": args
        ]

        let _: EmptyResponse = try await post(body: body)
    }

    // MARK: - Conversation APIs

    func getConversations(visitorId: String, sessionToken: String?, workspaceId: String) async throws -> [ConversationData] {
        var args: [String: Any] = [
            "visitorId": visitorId,
            "workspaceId": workspaceId
        ]
        if let sessionToken = sessionToken {
            args["sessionToken"] = sessionToken
        }
        let body: [String: Any] = [
            "path": "conversations:listByVisitor",
            "args": args
        ]

        let response: ConversationsResponse = try await post(body: body)
        return response.conversations
    }

    func getConversation(conversationId: String) async throws -> ConversationDetailData {
        let body: [String: Any] = [
            "path": "conversations:get",
            "args": ["conversationId": conversationId]
        ]

        return try await post(body: body)
    }

    func createConversation(visitorId: String, sessionToken: String?, workspaceId: String, initialMessage: String) async throws -> ConversationData {
        var args: [String: Any] = [
            "visitorId": visitorId,
            "workspaceId": workspaceId,
            "initialMessage": initialMessage
        ]
        if let sessionToken = sessionToken {
            args["sessionToken"] = sessionToken
        }
        let body: [String: Any] = [
            "path": "conversations:createForVisitor",
            "args": args
        ]

        return try await post(body: body)
    }

    func sendMessage(conversationId: String, content: String, visitorId: String, sessionToken: String?) async throws -> MessageData {
        var args: [String: Any] = [
            "conversationId": conversationId,
            "content": content,
            "senderId": visitorId,
            "senderType": "visitor",
            "visitorId": visitorId
        ]
        if let sessionToken = sessionToken {
            args["sessionToken"] = sessionToken
        }
        let body: [String: Any] = [
            "path": "messages:send",
            "args": args
        ]

        return try await post(body: body)
    }

    func getMessages(conversationId: String, visitorId: String?, sessionToken: String?) async throws -> [MessageData] {
        var args: [String: Any] = ["conversationId": conversationId]
        if let visitorId = visitorId {
            args["visitorId"] = visitorId
        }
        if let sessionToken = sessionToken {
            args["sessionToken"] = sessionToken
        }
        let body: [String: Any] = [
            "path": "messages:list",
            "args": args
        ]

        let response: MessagesResponse = try await post(body: body)
        return response.messages
    }

    // MARK: - Article APIs

    func getArticles(workspaceId: String) async throws -> [ArticleData] {
        let body: [String: Any] = [
            "path": "articles:listPublished",
            "args": ["workspaceId": workspaceId]
        ]

        let response: ArticlesResponse = try await post(body: body)
        return response.articles
    }

    func searchArticles(workspaceId: String, query: String) async throws -> [ArticleData] {
        let body: [String: Any] = [
            "path": "articles:search",
            "args": [
                "workspaceId": workspaceId,
                "query": query
            ]
        ]

        let response: ArticlesResponse = try await post(body: body)
        return response.articles
    }

    func getArticle(articleId: String) async throws -> ArticleData {
        let body: [String: Any] = [
            "path": "articles:get",
            "args": ["articleId": articleId]
        ]

        return try await post(body: body)
    }

    // MARK: - Carousel APIs

    func getCarousel(carouselId: String) async throws -> CarouselData {
        let body: [String: Any] = [
            "path": "carousels:get",
            "args": ["carouselId": carouselId]
        ]

        return try await post(body: body)
    }

    func trackCarouselInteraction(
        carouselId: String,
        visitorId: String,
        action: String,
        screenIndex: Int?
    ) async throws {
        var args: [String: Any] = [
            "carouselId": carouselId,
            "visitorId": visitorId,
            "action": action
        ]

        if let screenIndex = screenIndex {
            args["screenIndex"] = screenIndex
        }

        let body: [String: Any] = [
            "path": "carousels:trackInteraction",
            "args": args
        ]

        let _: EmptyResponse = try await post(body: body)
    }

    // MARK: - Push Notification APIs

    func registerPushToken(visitorId: String, token: String, platform: String) async throws {
        let body: [String: Any] = [
            "path": "pushTokens:register",
            "args": [
                "visitorId": visitorId,
                "token": token,
                "platform": platform
            ]
        ]

        let _: EmptyResponse = try await post(body: body)
    }

    // MARK: - HTTP

    private func post<T: Decodable>(body: [String: Any]) async throws -> T {
        guard let url = URL(string: "\(config.convexUrl)/api/mutation") else {
            throw OpencomError.invalidResponse
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw OpencomError.invalidResponse
        }

        if config.debug {
            print("[OpencomSDK] Response status: \(httpResponse.statusCode)")
            if let responseString = String(data: data, encoding: .utf8) {
                print("[OpencomSDK] Response body: \(responseString)")
            }
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            throw OpencomError.networkError(
                NSError(domain: "OpencomSDK", code: httpResponse.statusCode)
            )
        }

        return try decoder.decode(T.self, from: data)
    }
}

// MARK: - Response Types

struct BootSessionResponse: Codable {
    let visitor: BootVisitor
    let sessionToken: String
    let expiresAt: Double

    struct BootVisitor: Codable {
        let _id: String
    }
}

struct RefreshSessionResponse: Codable {
    let sessionToken: String
    let expiresAt: Double
}

struct EmptyResponse: Codable {}

struct ConversationsResponse: Codable {
    let conversations: [ConversationData]
}

struct MessagesResponse: Codable {
    let messages: [MessageData]
}

struct ArticlesResponse: Codable {
    let articles: [ArticleData]
}

// MARK: - Data Types

public struct ConversationData: Codable, Identifiable {
    public let id: String
    public let lastMessage: String?
    public let lastMessageAt: Double?
    public let unreadCount: Int
    public let createdAt: Double

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case lastMessage
        case lastMessageAt
        case unreadCount
        case createdAt
    }
}

public struct ConversationDetailData: Codable {
    public let id: String
    public let messages: [MessageData]
    public let createdAt: Double

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case messages
        case createdAt
    }
}

public struct MessageData: Codable, Identifiable {
    public let id: String
    public let conversationId: String
    public let senderId: String
    public let senderType: String
    public let content: String
    public let createdAt: Double

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case conversationId
        case senderId
        case senderType
        case content
        case createdAt
    }

    public var isFromVisitor: Bool {
        senderType == "visitor"
    }
}

public struct ArticleData: Codable, Identifiable {
    public let id: String
    public let title: String
    public let content: String
    public let slug: String
    public let summary: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case title
        case content
        case slug
        case summary
    }
}

public struct CarouselData: Codable, Identifiable {
    public let id: String
    public let name: String
    public let screens: [CarouselScreen]

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name
        case screens
    }
}

public struct CarouselScreen: Codable, Identifiable {
    public let id: String
    public let title: String?
    public let body: String?
    public let imageUrl: String?
    public let buttons: [CarouselButton]?
}

public struct CarouselButton: Codable {
    public let text: String
    public let action: String
    public let url: String?
    public let deepLink: String?
}

// MARK: - Device Info

struct DeviceInfo {
    let os: String
    let platform: String
    let deviceType: String
    let appVersion: String?

    static func current() -> DeviceInfo {
        DeviceInfo(
            os: "iOS \(UIDevice.current.systemVersion)",
            platform: "ios",
            deviceType: UIDevice.current.userInterfaceIdiom == .pad ? "tablet" : "mobile",
            appVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        )
    }

    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "os": os,
            "platform": platform,
            "deviceType": deviceType
        ]
        if let appVersion = appVersion {
            dict["appVersion"] = appVersion
        }
        return dict
    }
}

File: packages/android-sdk/sdk/src/main/kotlin/com/opencom/sdk/core/OpencomAPIClient.kt
package com.opencom.sdk.core

import com.opencom.sdk.OpencomConfig
import com.opencom.sdk.OpencomError
import com.opencom.sdk.OpencomUser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

internal class OpencomAPIClient(private val config: OpencomConfig) {

    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .apply {
            if (config.debug) {
                addInterceptor(HttpLoggingInterceptor().apply {
                    level = HttpLoggingInterceptor.Level.BODY
                })
            }
        }
        .build()

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    private val baseUrl: String
        get() = config.convexUrl.trimEnd('/')

    suspend fun bootSession(): BootSessionResponse = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("workspaceId", config.workspaceId)
        }

        val response = post("/api/sessions/boot", body)
        BootSessionResponse(
            visitorId = response.getJSONObject("visitor").getString("_id"),
            sessionToken = response.getString("sessionToken"),
            expiresAt = response.getLong("expiresAt")
        )
    }

    suspend fun refreshSession(sessionToken: String): RefreshSessionResponse = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("workspaceId", config.workspaceId)
            put("sessionToken", sessionToken)
        }

        val response = post("/api/sessions/refresh", body)
        RefreshSessionResponse(
            sessionToken = response.getString("sessionToken"),
            expiresAt = response.getLong("expiresAt")
        )
    }

    suspend fun revokeSession(sessionToken: String): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("workspaceId", config.workspaceId)
            put("sessionToken", sessionToken)
        }

        post("/api/sessions/revoke", body)
    }

    suspend fun identifyVisitor(
        visitorId: String,
        sessionToken: String?,
        user: OpencomUser
    ): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("visitorId", visitorId)
            put("workspaceId", config.workspaceId)
            sessionToken?.let { put("sessionToken", it) }
            user.userId?.let { put("userId", it) }
            user.email?.let { put("email", it) }
            user.name?.let { put("name", it) }
            user.company?.let { put("company", it) }
            if (user.customAttributes.isNotEmpty()) {
                put("customAttributes", JSONObject(user.customAttributes))
            }
        }

        post("/api/visitors/identify", body)
    }

    suspend fun trackEvent(
        visitorId: String,
        sessionToken: String?,
        name: String,
        properties: Map<String, Any>?,
        sessionId: String
    ): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("visitorId", visitorId)
            put("workspaceId", config.workspaceId)
            sessionToken?.let { put("sessionToken", it) }
            put("name", name)
            put("sessionId", sessionId)
            properties?.let { put("properties", JSONObject(it)) }
        }

        post("/api/events/track", body)
    }

    suspend fun heartbeat(visitorId: String, sessionToken: String?): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("visitorId", visitorId)
            put("workspaceId", config.workspaceId)
            sessionToken?.let { put("sessionToken", it) }
        }

        post("/api/visitors/heartbeat", body)
    }

    suspend fun endSession(visitorId: String, sessionId: String): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("visitorId", visitorId)
            put("sessionId", sessionId)
            put("workspaceId", config.workspaceId)
        }

        post("/api/sessions/end", body)
    }

    suspend fun registerPushToken(
        visitorId: String,
        token: String,
        platform: String = "android"
    ): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("visitorId", visitorId)
            put("workspaceId", config.workspaceId)
            put("token", token)
            put("platform", platform)
        }

        post("/api/push/register", body)
    }

    suspend fun getConversations(visitorId: String, sessionToken: String? = null): List<Conversation> = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("visitorId", visitorId)
            put("workspaceId", config.workspaceId)
            sessionToken?.let { put("sessionToken", it) }
        }

        val response = post("/api/conversations/list", body)
        val conversations = response.optJSONArray("conversations") ?: JSONArray()

        (0 until conversations.length()).map { i ->
            val conv = conversations.getJSONObject(i)
            Conversation(
                id = conv.getString("id"),
                title = conv.optString("title"),
                lastMessage = conv.optString("lastMessage"),
                lastMessageAt = conv.optLong("lastMessageAt"),
                unreadCount = conv.optInt("unreadCount", 0)
            )
        }
    }

    suspend fun getMessages(conversationId: String, visitorId: String? = null, sessionToken: String? = null): List<Message> = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("conversationId", conversationId)
            put("workspaceId", config.workspaceId)
            visitorId?.let { put("visitorId", it) }
            sessionToken?.let { put("sessionToken", it) }
        }

        val response = post("/api/messages/list", body)
        val messages = response.optJSONArray("messages") ?: JSONArray()

        (0 until messages.length()).map { i ->
            val msg = messages.getJSONObject(i)
            Message(
                id = msg.getString("id"),
                conversationId = msg.getString("conversationId"),
                content = msg.getString("content"),
                senderType = msg.getString("senderType"),
                senderName = msg.optString("senderName"),
                createdAt = msg.getLong("createdAt"),
                isRead = msg.optBoolean("isRead", false)
            )
        }
    }

    suspend fun sendMessage(
        conversationId: String,
        visitorId: String,
        content: String,
        sessionToken: String? = null
    ): Message = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("conversationId", conversationId)
            put("visitorId", visitorId)
            put("workspaceId", config.workspaceId)
            put("content", content)
            sessionToken?.let { put("sessionToken", it) }
        }

        val response = post("/api/messages/send", body)
        Message(
            id = response.getString("id"),
            conversationId = conversationId,
            content = content,
            senderType = "visitor",
            senderName = null,
            createdAt = response.getLong("createdAt"),
            isRead = false
        )
    }

    suspend fun createConversation(visitorId: String, sessionToken: String? = null): Conversation = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("visitorId", visitorId)
            put("workspaceId", config.workspaceId)
            sessionToken?.let { put("sessionToken", it) }
        }

        val response = post("/api/conversations/create", body)
        Conversation(
            id = response.getString("id"),
            title = null,
            lastMessage = null,
            lastMessageAt = System.currentTimeMillis(),
            unreadCount = 0
        )
    }

    suspend fun getArticles(): List<Article> = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("workspaceId", config.workspaceId)
        }

        val response = post("/api/articles/list", body)
        val articles = response.optJSONArray("articles") ?: JSONArray()

        (0 until articles.length()).map { i ->
            val article = articles.getJSONObject(i)
            Article(
                id = article.getString("id"),
                title = article.getString("title"),
                summary = article.optString("summary"),
                content = article.optString("content"),
                collectionId = article.optString("collectionId")
            )
        }
    }

    suspend fun searchArticles(query: String): List<Article> = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("workspaceId", config.workspaceId)
            put("query", query)
        }

        val response = post("/api/articles/search", body)
        val articles = response.optJSONArray("articles") ?: JSONArray()

        (0 until articles.length()).map { i ->
            val article = articles.getJSONObject(i)
            Article(
                id = article.getString("id"),
                title = article.getString("title"),
                summary = article.optString("summary"),
                content = article.optString("content"),
                collectionId = article.optString("collectionId")
            )
        }
    }

    suspend fun getCarousel(carouselId: String): Carousel? = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("workspaceId", config.workspaceId)
            put("carouselId", carouselId)
        }

        try {
            val response = post("/api/carousels/get", body)
            val screens = response.optJSONArray("screens") ?: JSONArray()

            Carousel(
                id = response.getString("id"),
                name = response.optString("name"),
                screens = (0 until screens.length()).map { i ->
                    val screen = screens.getJSONObject(i)
                    CarouselScreen(
                        id = screen.getString("id"),
                        title = screen.optString("title"),
                        body = screen.optString("body"),
                        imageUrl = screen.optString("imageUrl"),
                        buttonText = screen.optString("buttonText"),
                        buttonAction = screen.optString("buttonAction")
                    )
                }
            )
        } catch (e: Exception) {
            null
        }
    }

    suspend fun trackCarouselInteraction(
        visitorId: String,
        carouselId: String,
        action: String,
        screenIndex: Int
    ): Unit = withContext(Dispatchers.IO) {
        val body = JSONObject().apply {
            put("visitorId", visitorId)
            put("workspaceId", config.workspaceId)
            put("carouselId", carouselId)
            put("action", action)
            put("screenIndex", screenIndex)
        }

        post("/api/carousels/track", body)
    }

    private fun post(path: String, body: JSONObject): JSONObject {
        val request = Request.Builder()
            .url("$baseUrl$path")
            .post(body.toString().toRequestBody(jsonMediaType))
            .header("Content-Type", "application/json")
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            val errorBody = response.body?.string()
            throw OpencomError.ApiError(response.code, errorBody)
        }

        val responseBody = response.body?.string() ?: "{}"
        return try {
            JSONObject(responseBody)
        } catch (e: Exception) {
            throw OpencomError.ParseError(e)
        }
    }

    data class BootSessionResponse(
        val visitorId: String,
        val sessionToken: String,
        val expiresAt: Long
    )

    data class RefreshSessionResponse(
        val sessionToken: String,
        val expiresAt: Long
    )
}

data class Conversation(
    val id: String,
    val title: String?,
    val lastMessage: String?,
    val lastMessageAt: Long,
    val unreadCount: Int
)

data class Message(
    val id: String,
    val conversationId: String,
    val content: String,
    val senderType: String,
    val senderName: String?,
    val createdAt: Long,
    val isRead: Boolean
)

data class Article(
    val id: String,
    val title: String,
    val summary: String?,
    val content: String?,
    val collectionId: String?
)

data class Carousel(
    val id: String,
    val name: String?,
    val screens: List<CarouselScreen>
)

data class CarouselScreen(
    val id: String,
    val title: String?,
    val body: String?,
    val imageUrl: String?,
    val buttonText: String?,
    val buttonAction: String?
)
